src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+
+            if (capacity <= this.nodes.Count)
+            {
+                return;
+            }
+
+            this.nodes.Capacity = capacity;
+        }
+
+        /// <summary>
+        /// Sets the capacity to the actual number of items in the priority queue,
+        /// if that is less than 90 percent of current capacity.
+        /// </summary>
+        public void TrimExcess()
+        {
+            this.nodes.TrimExcess();
+        }
+
+        /// <summary>
+        /// Removes the node at the specified index.
+        /// </summary>
+        private void Remove(int indexOfNodeToRemove)
+        {
+            // The idea is to replace the specified node by the very last
+            // node and shorten the array by one.
+
+            var lastNodeIndex = this.GetLastNodeIndex();
+            var lastNode = this.nodes[lastNodeIndex];
+            this.nodes.RemoveAt(lastNodeIndex);
+
+            // In case we wanted to remove the node that was the last one,
+            // we are done.
+
+            if (indexOfNodeToRemove == lastNodeIndex)
+            {
+                return;
+            }
+
+            // Our last node was erased from the array and needs to be
+            // inserted again. Of course, we will overwrite the node we
+            // wanted to remove. After that operation, we will need
+            // to restore the heap property (in general).
+
+            var nodeToRemove = this.nodes[indexOfNodeToRemove];
+
+            var relation = this.Comparer.Compare(lastNode.priority, nodeToRemove.priority);
+            this.PutAt(lastNode, indexOfNodeToRemove);
+
+            if (relation < 0)
+            {
+                this.MoveUp(lastNode, indexOfNodeToRemove);
+            }
+            else
+            {
+                this.MoveDown(lastNode, indexOfNodeToRemove);
+            }
+        }
+
+        /// <summary>
+        /// Puts a node at the specified index.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void PutAt((TElement element, TPriority priority) node, int index)
+        {
+            this.nodes[index] = node;
+        }
+
+        /// <summary>
+        /// Gets the index of the last node in the heap.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetLastNodeIndex() => this.nodes.Count - 1;
+
+        /// <summary>
+        /// Gets the index of an element's parent.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetParentIndex(int index) => (index - 1) / Arity;
+
+        /// <summary>
+        /// Gets the index of the first child of an element.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetFirstChildIndex(int index) => Arity * index + 1;
```

Comment:
Similarly
```suggestion
        private int GetFirstChildIndex(int index) => (index + 1) << 2;
```

Reply:
I feel that these would change the function, `Arity * index + 1` vs `(index + 1) << 2`. Perhaps you meant `(index << 2) + 1`?

I'm looking at the generated IL for
```
public int CurrentOption(int index) => 4 * index + 1;
public int Alternative(int index) => (index << 2) + 1;
```

but not sure there is gain:

```
.method public hidebysig 
        instance int32 CurrentOption (
            int32 index
        ) cil managed 
    {
        // Method begins at RVA 0x2053
        // Code size 6 (0x6)
        .maxstack 8

        IL_0000: ldc.i4.4
        IL_0001: ldarg.1
        IL_0002: mul
        IL_0003: ldc.i4.1
        IL_0004: add
        IL_0005: ret
    } // end of method C::CurrentOption

    .method public hidebysig 
        instance int32 Alternative (
            int32 index
        ) cil managed 
    {
        // Method begins at RVA 0x205a
        // Code size 6 (0x6)
        .maxstack 8

        IL_0000: ldarg.1
        IL_0001: ldc.i4.2
        IL_0002: shl
        IL_0003: ldc.i4.1
        IL_0004: add
        IL_0005: ret
    } // end of method C::Alternative
```

Reply:
> but not sure there is gain

IL size should be the same as you highlighted (just with a shl instead of a mul and a different const value loaded accordingly), and the JIT should produce idential asm for both, e.g.
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYZWAVyiwhAeQAOGPhAEAKQcMEATGAgCUDALwA+BigYAqBus0MA1A3IBuLj3rmhDAIIAbDDCgDscgDcYZVcLbT1DEI0EBgAeWIZSHRt7GgBfIA===